File: Dynamic-Programming/1143-M-Longest-Common-Subsequence/LCS.c
int max(int a, int b) {
    return a>b?a:b;
}

int longestCommonSubsequence(char* text1, char* text2) {
    int t1 = strlen(text1);
    int t2 = strlen(text2);
    if (t1==0 || t2==0)
        return 0;

    int** OPT = malloc(sizeof(int *) * (t2+1));
    if (!OPT) return -909; // faled to allocate memory

    for (int i=0; i<t2+1; i++) {
        OPT[i] = calloc(sizeof(int), t1+1);
        if (!OPT[i]) {
            for (int k=0; k<i; k++) {
                free(OPT[k]);
            }
            free(OPT);
            return -909; // failed to allocate memory
        }
    }

    for (int i=1; i<t2+1; i++) {
        for (int j=1; j<t1+1; j++) {
            if (text1[j-1] == text2[i-1])
                OPT[i][j] = 1 + OPT[i-1][j-1];
            else
                OPT[i][j] = max(OPT[i-1][j], OPT[i][j-1]);
        }
    }
    int result = OPT[t2][t1];
    for (int i=0; i<t2+1; i++) {
        free(OPT[i]);
    }
    free(OPT);
    return result;
}

File: Dynamic-Programming/1143-M-Longest-Common-Subsequence/LCS.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        t1 = len(text1)
        t2 = len(text2)
        
        if t1 == 0 or t2 == 0:
            return 0

        mat = [[0] * (t2 + 1) for _ in range(t1 + 1)]

        for i in range(1, t1 + 1):
            for j in range(1, t2 + 1):
                if text1[i-1] == text2[j-1]:
                    mat[i][j] = 1 + mat[i-1][j-1]
                else:
                    mat[i][j] = max(mat[i-1][j], mat[i][j-1])
                    
        return mat[t1][t2]